pkg/resource/manager.go

```diff
@@ -125,12 +146,281 @@ func (m *manager) BatchCreateGenericResources(ctx context.Context, request *pb.B
+	if err != nil {
+		return nil, err
+	}
+
+	var genericResources []*pb.GenericResource
+	for _, hit := range searchResponse.Hits.Hits {
+		var genericResource pb.GenericResource
+		err = protojson.UnmarshalOptions{
+			DiscardUnknown: true,
+		}.Unmarshal(hit.Source, &genericResource)
+		if err != nil {
+			return nil, err
+		}
+
+		genericResources = append(genericResources, &genericResource)
+	}
+
+	return &pb.ListGenericResourcesResponse{
+		GenericResources: genericResources,
+		NextPageToken:    searchResponse.NextPageToken,
+	}, nil
+}
+
+// ListGenericResourceVersions handles the main logic for fetching versions associated with a generic resource.
+func (m *manager) ListGenericResourceVersions(ctx context.Context, request *pb.ListGenericResourceVersionsRequest) (*pb.ListGenericResourceVersionsResponse, error) {
+	return nil, nil
+}
+
+func (m *manager) GetGenericResource(ctx context.Context, resourceName string, resourceType pb.ResourceType) (*pb.GenericResource, error) {
+	log := m.logger.Named("GetGenericResource").With(zap.String("resource", resourceName))
+
+	response, err := m.esClient.Get(ctx, &esutil.GetRequest{
+		Index: m.indexManager.AliasName(genericResourcesDocumentKind, ""),
+		DocumentId: genericResourceId(&pb.GenericResource{
+			Name: resourceName,
+			Type: resourceType,
+		}),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if !response.Found {
+		log.Debug("generic resource not found")
+
+		return nil, nil
+	}
+
+	var resource pb.GenericResource
+	err = protojson.Unmarshal(response.Source, &resource)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resource, nil
+}
+
+// genericResourceVersionsFromOccurrence will create a map of generic versions, keyed by their IDs, from an occurrence.
+// if the occurrence is a build occurrence, generic resource versions will also be created for each built artifact.
+func genericResourceVersionsFromOccurrence(o *grafeas_proto.Occurrence) (map[string]*pb.GenericResourceVersion, error) {
+	result := make(map[string]*pb.GenericResourceVersion)
+
+	id, version, err := genericResourceVersionFromResourceUri(o.Resource.Uri)
+	if err != nil {
+		return nil, err
+	}
+
+	result[id] = version
+
+	if o.Kind == common_go_proto.NoteKind_BUILD {
+		details := o.Details.(*grafeas_proto.Occurrence_Build)
+
+		for _, artifact := range details.Build.Provenance.BuiltArtifacts {
+			artifactId, artifactVersion, err := genericResourceVersionFromResourceUri(artifact.Id)
+			if err != nil {
+				return nil, err
+			}
+
+			artifactVersion.Names = artifact.Names
+			artifactVersion.Created = o.CreateTime
+			result[artifactId] = artifactVersion
+		}
+	}
+
+	return result, nil
+}
+
+func genericResourceVersionFromResourceUri(resourceUri string) (string, *pb.GenericResourceVersion, error) {
+	uriParts, err := parseResourceUri(resourceUri)
+	if err != nil {
+		return "", nil, err
+	}
+
+	return fmt.Sprintf("%s:%s", uriParts.resourceType, resourceUri), &pb.GenericResourceVersion{Version: resourceUri}, nil
```

Comment:
The version id looks like a little strange for non-Docker resources: `GIT:git://github.com/rode/rode@foobar`

Reply:
agreed - I'll be making the previously discussed ID change for this today before we merge this.  it'll just be the fully qualified URI, nothing else.